boot/bootutil/src/encrypted_psa.c

```diff
@@ -0,0 +1,454 @@
+        return -3;
+    }
+
+    if (mbedtls_asn1_get_alg(p, end, &alg, &param) != 0) {
+        return -4;
+    }
+
+    if (alg.ASN1_CONTEXT_MEMBER(len) != sizeof(ec_pubkey_oid) - 1 ||
+        memcmp(alg.ASN1_CONTEXT_MEMBER(p), ec_pubkey_oid, sizeof(ec_pubkey_oid) - 1)) {
+        return -5;
+    }
+
+    if (mbedtls_asn1_get_tag(p, end, &len, MBEDTLS_ASN1_OCTET_STRING) != 0) {
+        return -6;
+    }
+
+    if (mbedtls_asn1_get_tag(p, end, &len, MBEDTLS_ASN1_OCTET_STRING) != 0) {
+        return -7;
+    }
+
+    if (len != PRIV_KEY_LEN) {
+        return -8;
+    }
+
+    memcpy(private_key, *p, PRIV_KEY_LEN);
+    return 0;
+}
+
+void bootutil_aes_ctr_init(bootutil_aes_ctr_context *ctx)
+{
+    psa_status_t psa_ret = psa_crypto_init();
+
+    (void)ctx;
+
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("AES init PSA crypto init failed %d", psa_ret);
+        assert(0);
+    }
+}
+
+#if defined(MCUBOOT_ENC_IMAGES)
+extern const struct bootutil_key bootutil_enc_key;
+/*
+ * Decrypt an encryption key TLV.
+ *
+ * @param buf An encryption TLV read from flash (build time fixed length)
+ * @param enckey An AES-128 or AES-256 key sized buffer to store to plain key.
+ */
+int
+boot_decrypt_key(const uint8_t *buf, uint8_t *enckey)
+{
+    uint8_t derived_key[BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE + BOOTUTIL_CRYPTO_SHA256_DIGEST_SIZE];
+    uint8_t *cp;
+    uint8_t *cpend;
+    uint8_t private_key[PRIV_KEY_LEN];
+    size_t len;
+    psa_status_t psa_ret = PSA_ERROR_BAD_STATE;
+    psa_status_t psa_cleanup_ret = PSA_ERROR_BAD_STATE;
+    psa_key_id_t kid;
+    psa_key_attributes_t kattr = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_derivation_operation_t key_do = PSA_KEY_DERIVATION_OPERATION_INIT;
+    psa_algorithm_t key_do_alg;
+    int rc = -1;
+
+    cp = (uint8_t *)bootutil_enc_key.key;
+    cpend = cp + *bootutil_enc_key.len;
+
+    /* The psa_cipher_decrypt needs initialization vector of proper length at
+     * the beginning of the input buffer.
+     */
+    uint8_t iv_and_key[PSA_CIPHER_IV_LENGTH(PSA_KEY_TYPE_AES, PSA_ALG_CTR) +
+                       BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE];
+
+    psa_ret = psa_crypto_init();
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("AES crypto init failed %d", psa_ret);
+        return -1;
+    }
+
+    /*
+     * Load the stored X25519 decryption private key
+     */
+    rc = parse_x25519_enckey(&cp, cpend, private_key);
+    if (rc) {
+        return rc;
+    }
+
+    psa_set_key_type(&kattr, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_MONTGOMERY));
+    psa_set_key_usage_flags(&kattr, PSA_KEY_USAGE_DERIVE);
+    psa_set_key_algorithm(&kattr, PSA_ALG_ECDH);
+
+    psa_ret = psa_import_key(&kattr, private_key, sizeof(private_key), &kid);
+    memset(private_key, 0, sizeof(private_key));
+    psa_reset_key_attributes(&kattr);
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("Built-in key import failed %d", psa_ret);
+        return -1;
```

Comment:
likewise here

Reply:
I agree, but I will leave as it is for now.
The change, as proposed here, I will do but I need to do that for all encryption sources at once and propagate it to calling code, so that new definition would be used there too.